src/transformers/generation_tf_utils.py

```diff
@@ -2260,6 +2305,511 @@ def sample(
+        ```"""
+
+        def flatten_beam_dim(tensor, batch_axis=0):
+            """Flattens the first two dimensions of a non-scalar array."""
+            # ignore scalars (e.g. cache index)
+            if tf.rank(tensor) == 0:
+                return tensor
+            return tf.reshape(
+                tensor,
+                tensor.shape[:batch_axis]
+                + [tensor.shape[batch_axis] * tensor.shape[batch_axis + 1]]
+                + tensor.shape[batch_axis + 2 :],
+            )
+
+        def unflatten_beam_dim(tensor, batch_size, num_beams, batch_axis=0):
+            """Unflattens the first, flat batch*beam dimension of a non-scalar array."""
+            # ignore scalars (e.g. cache index)
+            if tf.rank(tensor) == 0:
+                return tensor
+            return tf.reshape(
+                tensor, tensor.shape[:batch_axis] + [batch_size, num_beams] + tensor.shape[batch_axis + 1 :]
+            )
+
+        def gather_beams(nested, beam_indices, batch_axis=0):
+            """
+            Gathers the beam slices indexed by beam_indices into new beam array.
+            """
+
+            def gather_fn(tensor):
+                # ignore scalars (e.g. cache index)
+                if tf.rank(tensor) == 0:
+                    return tensor
+                else:
+                    if batch_axis > 0:
+                        # pushes all dimentions before the batch to the end, so we get (batch, beam_id, ...)
+                        perm = [axis for axis in range(tf.rank(tensor)) if axis >= batch_axis] + list(
+                            range(batch_axis)
+                        )
+                        tensor = tf.transpose(tensor, perm=perm)
+
+                    gathered_tensor = tf.gather(params=tensor, indices=beam_indices, axis=1, batch_dims=1)
+                    if batch_axis > 0:
+                        # transposes back to the original dimensions
+                        perm = [axis for axis in range(tf.rank(tensor)) if axis >= batch_axis] + list(
+                            range(batch_axis)
+                        )
+                        perm = tf.math.invert_permutation(perm)
+                        gathered_tensor = tf.transpose(gathered_tensor, perm=perm)
+
+                    return gathered_tensor
+
+            return tf.nest.map_structure(gather_fn, nested)
+
+        # 1. init beam_search values
+        logits_processor = logits_processor if logits_processor is not None else TFLogitsProcessorList()
+
+        # TODO (Joao): after the removal of the old beam_search, do not make an exception of the repetition penalty
+        repetition_processor = TFLogitsProcessorList(
+            [p for p in logits_processor if isinstance(p, TFRepetitionPenaltyLogitsProcessor)]
+        )
+        logits_processor = TFLogitsProcessorList(
+            [p for p in logits_processor if not isinstance(p, TFRepetitionPenaltyLogitsProcessor)]
+        )
```

Comment:
In the old TF beam search, the repetition penalty is applied before the softmax ([link for the softmax](https://github.com/huggingface/transformers/blob/main/src/transformers/generation_tf_utils.py#L939) -- you can see it is above it). So we need to split the processor list in two, and applied them at different places.

Reply:
Even for `beam_sample` I would advocated to not split the `logits_processor` and instead do this (super edge-case) breaking change

Reply:
I don't have a strong opinion here, so happy to do as suggested. However, if we do it, we will have to update some tests (which is the reason why I separated the processors).

Are you okay with updating tests that have a repetition penalty in them?